internal/pkg/cli/app_show.go

```diff
@@ -71,112 +43,120 @@ type ShowAppOpts struct {
 	*GlobalOpts
 }
 
-// Ask asks for fields that are required but not passed in.
-func (opts *ShowAppOpts) Ask() error {
-	// get project
-	if opts.ProjectName() == "" {
-		projectName, err := opts.selectProject()
+// Validate returns an error if the values provided by the user are invalid.
+func (o *ShowAppOpts) Validate() error {
+	if o.ProjectName() != "" {
+		_, err := o.storeSvc.GetProject(o.ProjectName())
 		if err != nil {
 			return err
 		}
-		opts.projectName = projectName
 	}
-
-	// get application of the project
-	if opts.appName == "" {
-		appName, err := opts.selectApplication()
+	if o.appName != "" {
+		_, err := o.storeSvc.GetApplication(o.ProjectName(), o.appName)
 		if err != nil {
 			return err
 		}
-		opts.appName = appName
 	}
 
 	return nil
 }
 
-// Validate returns an error if the user inputs are invalid.
-func (opts *ShowAppOpts) Validate() error {
-	proj, err := opts.storeSvc.GetProject(opts.ProjectName())
-	if err != nil {
-		return fmt.Errorf("getting project: %w", err)
+// Ask asks for fields that are required but not passed in.
+func (o *ShowAppOpts) Ask() error {
+	if err := o.askProject(); err != nil {
+		return err
 	}
-	opts.proj = proj
-
-	return nil
+	return o.askAppName()
 }
 
 // Execute shows the applications through the prompt.
-func (opts *ShowAppOpts) Execute() error {
-	if opts.appName != "" {
-		if err := opts.retrieveData(); err != nil {
-			return err
-		}
-	} else {
-		opts.app = serializedWebApp{}
+func (o *ShowAppOpts) Execute() error {
+	app, err := o.retrieveData()
+	if err != nil {
+		return err
 	}
 
-	if opts.ShouldOutputJSON {
-		data, err := opts.jsonOutput()
+	if o.shouldOutputJSON {
+		data, err := o.jsonOutput(app)
 		if err != nil {
 			return err
 		}
-		fmt.Fprintf(opts.w, data)
+		fmt.Fprintf(o.w, data)
 	} else {
-		opts.humanOutput()
+		o.humanOutput(app)
 	}
 
 	return nil
 }
 
-func (opts *ShowAppOpts) retrieveData() error {
-	app, err := opts.storeSvc.GetApplication(opts.ProjectName(), opts.appName)
+func (o *ShowAppOpts) retrieveData() (*describe.WebApp, error) {
```

Comment:
I wonder if this method should belong to the `describer` instead since it returns a struct from the same package `describe.WebApp`? Same for `humanOutput` and `jsonOutput`

Here is what I'm thinking:
```go
webApp := o.describer.Describe()

s, err := webApp.HumanString() // returns the string from `func (o *ShowAppOpts) humanOutput(app *describe.WebApp)`
s, err := webApp.JSONString() // returns the string from `func (o *ShowAppOpts) jsonOutput(app *describe.WebApp) (string, error)`
```

The main benefit of this approach is that we can take dependencies on `HumanStringer` and `JSONStringer` interfaces in the future to describe different type of applications without changing the `app show` command significantly.